src/rebar_compiler_erl.erl

```diff
@@ -25,8 +25,16 @@ context(AppInfo) ->
     RebarOpts = rebar_app_info:opts(AppInfo),
     ErlOpts = rebar_opts:erl_opts(RebarOpts),
     ErlOptIncludes = proplists:get_all_values(i, ErlOpts),
-    InclDirs = lists:map(fun(Incl) -> filename:absname(Incl) end, ErlOptIncludes),
-    AbsIncl = [filename:join([OutDir, "include"]) | InclDirs],
+    AbsIncl = [filename:join(OutDir, "include") | % standard include path
+               %% includes specified by erl_opts
+               lists:map(fun(Incl) -> filename:absname(Incl) end, ErlOptIncludes)] ++
+              %% all source directories are valid, and might also be recursive
+              lists:append([
+                  find_recursive_incl(OutDir, Src, RebarOpts) ||
+                  Src <- rebar_dir:all_src_dirs(RebarOpts, ["src"], [])
+              ]) ++
+              %% top-level dir for legacy stuff
+              [OutDir],
```

Comment:
For reviews' sake: I got the list for these paths from line 242-246. Since we were passing them to the compiler, they had to be right already.